### PR TITLE
MBP's context in a output port. 

### PR DESCRIPTION
**Pull request by siyuanfeng-tri**
I think there are a bunch of use cases for a system to take kinematics as input. The most common solution that I am aware of is for these systems to declare a state input, then with either a mutable member field Context or allocating a Context on every computation to do the kinematics. Neither is pretty. So I made the kinematics_calculator system, and frame_visualizer is a demo application that I revived from attic. 
@amcastro-tri @sherm1 @RussTedrake @edrumwri what do you guys think?

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/12662)

### CODE REVIEW

**Comment by sherm1** [comment]
@edrumwri what I meant was literally a `mutable Context<double> context_;` member in the receiving LeafSystem. That could be made available to anyone who needs it including output port Calc methods -- wherever it's needed it would be preceded by reading the state input port and stuffing the values in.

A somewhat better alternative would be to put that local Context into a cache entry in the main Context. That cache entry's Calc() method would read the port and fill in the Context. That would be identical to what @siyuanfeng-tri did (since his output port is associated with a cache entry). That has the advantages (a) it is accessible to anyone who has access to the main Context, (b) it doesn't require a mutable main Context (but is itself mutable if that's needed), and (c) it knows whether it is up to date w.r.t. the state input. Maybe that's too much to put in every LeafSystem, thus justifying the implementation in a separate System as Siyuan proposes?

I think that's a nice way to

**Comment by sherm1** [comment]
Yes, I think the cached Context is the tidiest but keeping mutable temp variables around to avoid repeated reallocation is rather ... traditional. You just have to promise always to set them before use and not to dole them out to multiple threads.

BTW Siyuan's KinematicCalculator achieves exactly the same thing as a locally-cached Context -- I think it is a matter of style whether to use that in a Diagram vs. cache the Context locally.

**Comment by jwnimmer-tri** [comment]
(I've added a DNR label to take it off of platform reviewers' radar.)

**Comment by jwnimmer-tri** [comment]
FWIW This seems like the kind of thing we should prove out in Anzu first to work out the kinks, before we share with the world.

**Comment by siyuanfeng-tri** [comment]
@jwnimmer-tri any suggestion for usage in anzu? we do really do system at all in anzu.

**Comment by jwnimmer-tri** [comment]
Well, if it's not used in Anzu, where are you going to use it?  If we have no uses of it, we should not put it onto master in the first place.

If you're going to use it in a Drake example, you could place it in that example's folder for now, until it proves its worth.  But maybe that example should be developed in Anzu, too, to go faster.

**Comment by siyuanfeng-tri** [comment]
i have used the kin_cal + frame_viz with a sim in drake, and didn't observe any obvious issue or performance drop. 
i could start using it in my drake robot_bridge test branch.

**Comment by siyuanfeng-tri** [comment]
> Well, if it's not used in Anzu, where are you going to use it? 
well, anzu just dont do systems.

i do use it locally in some other multi k line change dev branch. and i feel it is a useful thing to have. also i think we can replace all of the PoseBundle stuff in systems/rendering with this.

**Comment by jwnimmer-tri** [comment]
The capability might be useful, but what if the design is broken?  Will we know until we start building applications with it?  Putting it in the `plant` folder suggests that it's ready for prime time, but we have no evidence of that.

**Comment by jwnimmer-tri** [comment]
Playing fast and loose with subsystem contexts is very finicky.  If the plant system pointer used to interrogate the kinematic contexts does not match the context, bad things could happen.  Cross-wiring System pointer aliases within a Diagram, without evaluating through ports, is a dangerous game to play and we need to have a convincing proof why it is safe.

**Comment by jwnimmer-tri** [comment]
> I totally agree, and it is a safety check i have been asking around for a while.

We already (recently) check that System and Context are matched (#12551).  What that kind of check does not catch is if there are other parameters or inputs to the upstream MbP which are not reflected in the **default** context that this computation uses, either from the Diagram wiring that the MbP is part of, or Context parameters that influence the computation.  Not all of those cases (do or even can) fail-fast when the user makes a mistake.

> What's a better alternative?

Problem statement unclear.  If you want some design help, let's take it offline by looking at your actual application code.

**Comment by sherm1** [comment]
I'm not as worried about this as I think you are @jwnimmer-tri as long as it is limited just to kinematics in MBP. For that all that's needed is q and v (currently), no ports. However if the MBP had parameters (say lengths of links) then this approach would result in the parameters having their default values. Not a problem currently.

**Comment by siyuanfeng-tri** [comment]
All this pr tries to solve is to have a central place to query for generic FK related questions, as opposed to every subsystem aliasing some MBP, allocating and managing their own contexts, and repeating similar computations, which is what we do now. InverseDynamics under systems/controllers is one simple example. The humanoid stuff also needed to do a lot of the FK related computations. This is a real and common use case. 

> What that kind of check does not catch is if there are other parameters or inputs to the upstream MbP which are not reflected in the default context that this computation uses, either from the Diagram wiring that the MbP is part of, or Context parameters that influence the computation.

I agree. But unless I can ask all FK questions through MBP's outputs, or having the one and only MBP expose its context in an output, there is no way for anyone to query these correctly. All of the current controller designs are flawed in the same way.

**Comment by jwnimmer-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/12662#-M-jdNkDDcHIBaMsjFKs:-M-jmf724bKbhUqWd47x:b-gyfdfa) (no related file):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Late to the party (out for vacation). Maybe a dumb question. What about having an MBP output port for all frame poses? then anybody who need that computation would just connect to that port. would that work?
</blockquote></details>

I think not.

There are plenty of kinematic methods that take an argument besides Context or a FrameId, e.g., `GetPositionsAndVelocities(..., ModelInstanceIndex)` or `CalcRelativeTransform(..., const Frame<T>& frame_F, const Frame<T>& frame_G)` or `CalcPointsPositions(...)`.

Basically the MbP API is wide, and so it's not plausible to condense that all into a sane number of unparameterized output ports.  Plus anytime you group more than one fact into a single output port, you're going to induce computation of unrelated facts just to get the one you want.

---

**Comment by sherm1** [comment]
Also, FWIW passing along a full live context defeats any attempt to track dependencies through a Diagram. It is a huge loophole in my view -- likely to prevent differentiation and most analyses. Note that @siyuanfeng-tri's proposal does not have that effect because it just uses a local Context object as a place to put pre-calculated kinematic results. That's not the live Context being used to advance the Diagram trajectory.

I like Alejandro's idea of adding a kinematics output port though. MBP already has an assortment of full and model instance-specific outputs:
image

It wouldn't be unreasonable to add more if that provided a clean solution to a common problem.

**Comment by siyuanfeng-tri** [comment]
what about exposing PositionKinematicsCache and VelocityKinematicsCache and maybe AccelerationKinematicsCache?

**Comment by siyuanfeng-tri** [comment]
> What about having an MBP output port for all frame poses? then anybody who need that computation would just connect to that port. would that work?

This only solves the position part of the problem. I think any function of x is a fair question to ask. Like wheres is my center of mass, or what's my inertia matrix, etc.

**Comment by jwnimmer-tri** [comment]
@siyuanfeng-tri This has been open for almost 6 months with no changes.  I know the problem still exists, but is this topic still relevant to actively pursue?  A lot of the discussion in this PR seems to be more like an issue thread, instead of talking about the code in this PR.  Would it make sense to close the PR for now (and maybe open an issue?)

**Comment by jwnimmer-tri** [comment]
Closing after ~1 year with no motion.
